### ISSUE TITLE
How to extract method, variable etc. using this plugin?

### ISSUE THREAD

**Issue by LintaoAmons** [ray-x/go.nvim#207]
I saw in the README, that I can do the refactor stuff using your plugin.

```
Code refactor made easy: GoFixPlural, FixStruct, FixSwitch, Add comment, IfErr, ModTidy, GoGet, extract function/block with codeactions... Most of the tools are built on top of treesitter AST or go AST. It is fast and accurate.
```

But I didn't found the command to do the refactor... So could you please tell me how to achieve that?

**Comment by alphatroya** [comment]
![Снимок экрана 2022-09-01 в 22 11 26](https://user-images.githubusercontent.com/4927633/187993859-cd4ee6bd-9403-4a62-8d43-9eae876eddd4.png)
[vim-go](https://github.com/fatih/vim-go) has a :GoFreevars command. It's a very useful during large method extraction. As I see it, this plugin does not yet have a similar command.

**Comment by ray-x** [comment]
Those are handled by codeaction with gopls.

<img width="618" alt="image" src="https://user-images.githubusercontent.com/1681295/188039321-b95acf61-b139-44ab-a7d5-b5d5200bde53.png">

**Comment by LintaoAmons** [comment]
Sorry, I quite new to lsp stuff... How to trigger the codeaction like your picture?

**Comment by ray-x** [comment]
Depends on your setup
`GoCodeAction` provided by go.nvim
Or other lsp plugins that have code action supported. Above image is captured with navigator.lua which hook to lsp.

**Comment by LintaoAmons** [comment]
I have add `go.nvim` and use `mason` to installed all options it provided that start with `go.*`

But when I use `GoCodeAction` it told me that there's no `action` available?

![go-code-action](https://user-images.githubusercontent.com/95092244/188073171-b9f122fc-2589-48e4-8418-26b95997bb2d.gif)

**Comment by etrnal70** [comment]
It works for me. You need to use code action on ranges (visual block on the conditional block)

**Comment by ray-x** [comment]
The visual mode only works with keybindings.
I update the command so it will work in command mode as well
Please update the plugin and use `GoCodeAction` in visual mode.

**Comment by ray-x** [comment]
@LintaoAmons Can you confirm if it works for you?

**Comment by LintaoAmons** [comment]
I don't know where I went wrong? 

I update the `go.nvim` plugin, and be able to use `GoCodeAction` command with selected range .

But still `No code action available`...

Don't know why

**Comment by gonzaloserrano** [comment]
Same here.

**Comment by alphatroya** [comment]
I've done the same behaviour by using [refactoring.nvim](https://github.com/ThePrimeagen/refactoring.nvim) plugin. 
If you configure the [null-ls.nvim](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTINS.md#refactoring) plugin buildin, all refactoring suggestion will be passed to LSP code action menu.

**Comment by ray-x** [comment]
not every piece of go code has range codeaction
But you should see something like this if code action available:
![image](https://user-images.githubusercontent.com/1681295/202127679-18092478-2e76-49fe-a4e0-83b7d8256c1e.png)

**Comment by gonzaloserrano** [comment]
I've migrated to my own config from lunarvim (you can see how here https://github.com/gonzaloserrano/dotfiles/tree/master/nvim) and now `:GoCodeAction` works!

Is there any way to map one of the actions? I.e I use extract func a lot, so I'd like to do `:GoCodeAction 2` or similar.

**Comment by ray-x** [comment]
If you enabled `lsp_keymaps`, the default mapping:
```lua
    buf_set_keymap('n', '<space>ca', "<cmd>lua require('go.codeaction').run_code_action()<CR>", opts)
    buf_set_keymap('v', '<space>ca', "<cmd>lua require('go.codeaction').run_range_code_action()<CR>", opts)

```
You can set mapping key of your own. Please check lsp.lua on how it is mapped.

**Comment by gonzaloserrano** [comment]
Sorry, what I mean is: once I map the code action, I'm given several options, i.e 
![CleanShot 2022-11-22 at 11 33 03](https://user-images.githubusercontent.com/349328/203291721-3b9a8a31-59f4-4dd3-bc2a-a3d32d512c96.png)
Is there any way to map the options?

**Comment by ray-x** [comment]
yes and no.
As there are lots of methods provided by gopls. I did not map those. I can do that later.
ATM, in UI menu, you can press 1~9 to navigate to the menu item.

**Comment by phonzia** [comment]
Could you provide a example of lsp and code action map configuration?

**Comment by testinfected** [comment]
Thanks for the great plugin @ray-x 🥇 

Any idea why I only get blame code when using `:GoCodeAction` but I get more options with `vim.lsp.buf.code_action()`, such as the option to extract function?

I am in visual mode, lines selected. I'm using mason with `lsp_cfg = false` and 

```
    local cfg = require('go.lsp').config()
    require('lspconfig').gopls.setup(cfg)
```

as suggested in the docs.

Thanks 🙏

**Comment by ray-x** [comment]
In Visual mode you should see this:
<img width="421" alt="image" src="https://github.com/ray-x/go.nvim/assets/1681295/148040df-f642-4a84-829b-46ccf75c98d4">
Please use `:'<,'>GoCodeAction`

**Comment by testinfected** [comment]
I'm only seeing Blame line, which comes from null-ls

<img width="1029" alt="Screenshot 2023-08-22 at 00 07 31" src="https://github.com/ray-x/go.nvim/assets/128804/9f98c7fb-a80b-451e-91c1-3df2b6e1e9ad">

**Comment by ray-x** [comment]
Look to me when you call `GoCodeAction`, the gopls was not attached. Could you check your Gopls status when error happens?
You can run `:LspInfo` to get info.

**Comment by ray-x** [comment]
It looks good.
Could you try to disable null-ls and try it again? I have a feeling is gopls is too slow in your project.

**Comment by testinfected** [comment]
What I don't understand is that this mapping works:

vim.key.set('v', '<leader>ca', vim.lsp.buf.code_action, { desc = "Range code action" })

<img width="1220" alt="Screenshot 2023-08-23 at 09 27 17" src="https://github.com/ray-x/go.nvim/assets/128804/b0d47ccf-5ca4-4380-85c8-5dc2d30d66af">

**Comment by testinfected** [comment]
I turned on the verbose option, and I get this message each time it loads the go file:

<img width="420" alt="Screenshot 2023-08-23 at 09 34 05" src="https://github.com/ray-x/go.nvim/assets/128804/88d2667a-f37e-4234-b18d-28870cd1b733">

My goals is 0.13.2:
<img width="409" alt="Screenshot 2023-08-23 at 09 36 15" src="https://github.com/ray-x/go.nvim/assets/128804/3c1ba51a-c9e8-40ba-92b5-1444150b4512">

**Comment by testinfected** [comment]
Something is wrong cause I get No code actions even if the playground go project using the provided `init.lua` config

**Comment by testinfected** [comment]
Ok good news, finally figured it out. It turns out the visual selection is pretty picky. For instance it does not work with visual line selection. Also you've got to be pretty precise with your selection (don't include an extra CR for instance) otherwise it won't work.

@ray-x Do you see anyway to make it more tolerant with the selection (i.e. support visual line mode, tolerate extra spaces and newlines, etc.)
<img width="899" alt="Screenshot 2023-08-23 at 16 09 44" src="https://github.com/ray-x/go.nvim/assets/128804/0f214024-cee6-45e2-97d8-e1b8547432df">

**Comment by gonzaloserrano** [comment]
Yeah, I just noticed that too.

A selection with visual line does not work:
<img width="1027" alt="CleanShot 2023-10-06 at 16 57 57@2x" src="https://github.com/ray-x/go.nvim/assets/349328/aefbe862-5aae-4617-a6ac-f25ff769f018">

But if I do block selection and just get the characters, then it works:

<img width="1010" alt="CleanShot 2023-10-06 at 16 58 11@2x" src="https://github.com/ray-x/go.nvim/assets/349328/5c2f49b8-a114-467e-bcce-9a3e3fe324ac">

**Comment by ray-x** [comment]
Yes, there is something incorrect with recent release of neovim. This one should have been fix with updates listed above

**Comment by gonzaloserrano** [comment]
hi, just fyi what I described still happens with nvim v0.10.1 and go.nvim master (`9743cc3`), i.e line selection does not trigger "extract function" but block selection of the same line does.
